File: python/agentseal/toxic_flows.py
# agentseal/toxic_flows.py
"""
Lightweight toxic flow detection - static analysis only (no MCP runtime).

Classifies MCP servers by capability labels based on known package names
and server name heuristics. Detects dangerous combinations of capabilities
across servers that could enable data exfiltration, remote code execution,
or data destruction.

Wave 2 scope: classification from config only (names + args).
Phase 2 upgrade: replace with actual tool descriptions from list_tools().
"""

import re
from dataclasses import dataclass, field
from typing import Optional


# ═══════════════════════════════════════════════════════════════════════
# Capability Labels
# ═══════════════════════════════════════════════════════════════════════

LABEL_PUBLIC_SINK = "public_sink"       # sends data externally
LABEL_DESTRUCTIVE = "destructive"       # modifies/deletes data
LABEL_UNTRUSTED = "untrusted_content"   # fetches external data
LABEL_PRIVATE = "private_data"          # reads sensitive data

ALL_LABELS = {LABEL_PUBLIC_SINK, LABEL_DESTRUCTIVE, LABEL_UNTRUSTED, LABEL_PRIVATE}


# ═══════════════════════════════════════════════════════════════════════
# Known Server Classifications
# ═══════════════════════════════════════════════════════════════════════

# Curated mapping of well-known MCP server packages to their capability labels.
KNOWN_SERVER_LABELS: dict[str, set[str]] = {
    # Filesystem
    "filesystem": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "fs": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    # Communication / sinks
    "slack": {LABEL_PUBLIC_SINK},
    "discord": {LABEL_PUBLIC_SINK},
    "email": {LABEL_PUBLIC_SINK},
    "gmail": {LABEL_PUBLIC_SINK},
    "smtp": {LABEL_PUBLIC_SINK},
    "sendgrid": {LABEL_PUBLIC_SINK},
    "twilio": {LABEL_PUBLIC_SINK},
    "telegram": {LABEL_PUBLIC_SINK},
    "teams": {LABEL_PUBLIC_SINK},
    "webhook": {LABEL_PUBLIC_SINK},
    # Code/project platforms
    "github": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    "gitlab": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    "bitbucket": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    "linear": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    "jira": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    "notion": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    "asana": {LABEL_PUBLIC_SINK, LABEL_PRIVATE},
    # Databases
    "postgres": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "postgresql": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "mysql": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "sqlite": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "mongo": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "mongodb": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "redis": {LABEL_PRIVATE, LABEL_DESTRUCTIVE},
    "supabase": {LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK},
    # Web / external content
    "fetch": {LABEL_UNTRUSTED},
    "puppeteer": {LABEL_UNTRUSTED},
    "playwright": {LABEL_UNTRUSTED},
    "browser": {LABEL_UNTRUSTED},
    "brave-search": {LABEL_UNTRUSTED},
    "tavily": {LABEL_UNTRUSTED},
    "web-search": {LABEL_UNTRUSTED},
    "scraper": {LABEL_UNTRUSTED},
    "crawl": {LABEL_UNTRUSTED},
    # Infrastructure
    "aws": {LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK},
    "gcp": {LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK},
    "azure": {LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK},
    "docker": {LABEL_DESTRUCTIVE},
    "kubernetes": {LABEL_DESTRUCTIVE},
    "k8s": {LABEL_DESTRUCTIVE},
    "terraform": {LABEL_DESTRUCTIVE},
    # Execution
    "shell": {LABEL_DESTRUCTIVE, LABEL_UNTRUSTED},
    "terminal": {LABEL_DESTRUCTIVE, LABEL_UNTRUSTED},
    "exec": {LABEL_DESTRUCTIVE},
    "code-runner": {LABEL_DESTRUCTIVE},
    "sandbox": {LABEL_DESTRUCTIVE},
    # Memory / state
    "memory": {LABEL_PRIVATE},
    "knowledge": {LABEL_PRIVATE},
    "vector": {LABEL_PRIVATE},
    # Monitoring
    "sentry": {LABEL_PRIVATE},
    "datadog": {LABEL_PRIVATE},
    "grafana": {LABEL_PRIVATE},
    # Storage
    "s3": {LABEL_PRIVATE, LABEL_PUBLIC_SINK, LABEL_DESTRUCTIVE},
    "gcs": {LABEL_PRIVATE, LABEL_PUBLIC_SINK, LABEL_DESTRUCTIVE},
    "drive": {LABEL_PRIVATE, LABEL_PUBLIC_SINK},
    "dropbox": {LABEL_PRIVATE, LABEL_PUBLIC_SINK},
}


# Heuristic patterns for servers not in the known list.
_NAME_HEURISTICS: list[tuple[re.Pattern, set[str]]] = [
    (re.compile(r"(?:file|fs|disk)", re.I), {LABEL_PRIVATE, LABEL_DESTRUCTIVE}),
    (re.compile(r"(?:mail|email|smtp)", re.I), {LABEL_PUBLIC_SINK}),
    (re.compile(r"(?:http|fetch|web|browser|scrape|crawl)", re.I), {LABEL_UNTRUSTED}),
    (re.compile(r"(?:db|sql|database|mongo|redis)", re.I), {LABEL_PRIVATE}),
    (re.compile(r"(?:exec|shell|command|terminal|run)", re.I), {LABEL_DESTRUCTIVE}),
    (re.compile(r"(?:slack|discord|teams|telegram|chat)", re.I), {LABEL_PUBLIC_SINK}),
    (re.compile(r"(?:github|gitlab|bitbucket|jira|linear)", re.I), {LABEL_PUBLIC_SINK, LABEL_PRIVATE}),
    (re.compile(r"(?:aws|gcp|azure|cloud)", re.I), {LABEL_PRIVATE, LABEL_DESTRUCTIVE}),
    (re.compile(r"(?:docker|k8s|kubernetes|terraform)", re.I), {LABEL_DESTRUCTIVE}),
    (re.compile(r"(?:s3|gcs|storage|drive|dropbox)", re.I), {LABEL_PRIVATE, LABEL_PUBLIC_SINK}),
]


def classify_server(server: dict) -> set[str]:
    """Classify an MCP server by its capability labels.

    Checks known package names first, then falls back to name heuristics.
    """
    name = server.get("name", "").lower().strip()
    command = server.get("command", "").lower()
    args_str = " ".join(str(a) for a in server.get("args", []) if isinstance(a, str)).lower()

    # Check known server names (exact match)
    if name in KNOWN_SERVER_LABELS:
        return set(KNOWN_SERVER_LABELS[name])

    # Check if any known name appears in the server name
    for known, labels in KNOWN_SERVER_LABELS.items():
        if known in name:
            return set(labels)

    # Check if any known name appears in command or args (package names)
    search_text = f"{command} {args_str}"
    for known, labels in KNOWN_SERVER_LABELS.items():
        if known in search_text:
            return set(labels)

    # Fall back to heuristic patterns
    labels: set[str] = set()
    for pattern, heuristic_labels in _NAME_HEURISTICS:
        if pattern.search(name) or pattern.search(command) or pattern.search(args_str):
            labels |= heuristic_labels

    return labels


# ═══════════════════════════════════════════════════════════════════════
# Dangerous Combinations
# ═══════════════════════════════════════════════════════════════════════

@dataclass
class ToxicFlow:
    """A detected dangerous combination of server capabilities."""
    risk_level: str  # "high", "medium"
    risk_type: str   # "data_exfiltration", "remote_code_execution", "data_destruction", "full_chain"
    title: str
    description: str
    servers_involved: list[str]
    labels_involved: list[str]
    remediation: str


def _detect_combos(
    server_labels: dict[str, set[str]],
) -> list[ToxicFlow]:
    """Detect dangerous capability combinations across servers."""
    flows: list[ToxicFlow] = []

    # Collect all labels across all servers
    all_labels: set[str] = set()
    for labels in server_labels.values():
        all_labels |= labels

    # Servers by label
    by_label: dict[str, list[str]] = {}
    for name, labels in server_labels.items():
        for label in labels:
            by_label.setdefault(label, []).append(name)

    has_private = LABEL_PRIVATE in all_labels
    has_sink = LABEL_PUBLIC_SINK in all_labels
    has_untrusted = LABEL_UNTRUSTED in all_labels
    has_destructive = LABEL_DESTRUCTIVE in all_labels

    # Full chain: untrusted + private + sink (highest risk)
    if has_untrusted and has_private and has_sink:
        flows.append(ToxicFlow(
            risk_level="high",
            risk_type="full_chain",
            title="Full attack chain detected",
            description=(
                "This agent can fetch external content, read private data, "
                "and send data externally. An attacker could inject instructions "
                "via fetched content, read sensitive files, and exfiltrate them."
            ),
            servers_involved=sorted(set(
                by_label.get(LABEL_UNTRUSTED, []) +
                by_label.get(LABEL_PRIVATE, []) +
                by_label.get(LABEL_PUBLIC_SINK, [])
            )),
            labels_involved=[LABEL_UNTRUSTED, LABEL_PRIVATE, LABEL_PUBLIC_SINK],
            remediation=(
                "Scope filesystem access to non-sensitive directories. "
                "Remove or restrict external communication servers."
            ),
        ))
        return flows  # Full chain subsumes the individual combos

    # Data exfiltration: private + sink
    if has_private and has_sink:
        flows.append(ToxicFlow(
            risk_level="high",
            risk_type="data_exfiltration",
            title="Data exfiltration path detected",
            description=(
                "This agent can read private data and send it externally. "
                "A prompt injection could instruct the agent to read sensitive "
                "files and leak them via an external service."
            ),
            servers_involved=sorted(set(
                by_label.get(LABEL_PRIVATE, []) +
                by_label.get(LABEL_PUBLIC_SINK, [])
            )),
            labels_involved=[LABEL_PRIVATE, LABEL_PUBLIC_SINK],
            remediation=(
                "Scope filesystem access to non-sensitive directories only. "
                "Review which external services truly need write access."
            ),
        ))

    # Remote code execution: untrusted + destructive
    if has_untrusted and has_destructive:
        flows.append(ToxicFlow(
            risk_level="high",
            risk_type="remote_code_execution",
            title="Remote code execution path detected",
            description=(
                "This agent can fetch external content and execute destructive "
                "operations. Fetched content could contain malicious instructions "
                "that modify files, execute commands, or alter databases."
            ),
            servers_involved=sorted(set(
                by_label.get(LABEL_UNTRUSTED, []) +
                by_label.get(LABEL_DESTRUCTIVE, [])
            )),
            labels_involved=[LABEL_UNTRUSTED, LABEL_DESTRUCTIVE],
            remediation=(
                "Add confirmation steps before destructive operations. "
                "Restrict or sandbox the execution server."
            ),
        ))

    # Data destruction: private + destructive (from different servers)
    if has_private and has_destructive:
        private_servers = set(by_label.get(LABEL_PRIVATE, []))
        destructive_servers = set(by_label.get(LABEL_DESTRUCTIVE, []))
        # Only flag if the capability spans multiple servers
        # (a single server like filesystem inherently has both)
        if private_servers != destructive_servers:
            flows.append(ToxicFlow(
                risk_level="medium",
                risk_type="data_destruction",
                title="Data destruction path detected",
                description=(
                    "This agent can read private data from one source and "
                    "perform destructive operations on another. This could "
                    "lead to data corruption or deletion."
                ),
                servers_involved=sorted(private_servers | destructive_servers),
                labels_involved=[LABEL_PRIVATE, LABEL_DESTRUCTIVE],
                remediation=(
                    "Review whether both data read and write capabilities "
                    "are necessary. Consider read-only access where possible."
                ),
            ))

    return flows


def analyze_toxic_flows(servers: list[dict]) -> list[ToxicFlow]:
    """Analyze MCP servers for dangerous capability combinations.

    Args:
        servers: List of MCP server config dicts (as returned by scan_machine).

    Returns:
        List of detected toxic flows (empty if no dangerous combos found).
    """
    if len(servers) < 2:
        return []  # Need at least 2 servers for a cross-server flow

    server_labels: dict[str, set[str]] = {}
    for srv in servers:
        name = srv.get("name", "unknown")
        labels = classify_server(srv)
        if labels:
            server_labels[name] = labels

    if not server_labels:
        return []

    return _detect_combos(server_labels)

File: js/src/toxic-flows.ts
/**
 * Toxic flow detection — static analysis of MCP server capability combinations.
 *
 * Classifies MCP servers by capability labels and detects dangerous
 * combinations that could enable data exfiltration, remote code execution,
 * or data destruction.
 *
 * Port of Python agentseal/toxic_flows.py — static analysis only.
 */

import type { ToxicFlowResult } from "./guard-models.js";

// ═══════════════════════════════════════════════════════════════════════
// Capability Labels
// ═══════════════════════════════════════════════════════════════════════

export const LABEL_PUBLIC_SINK = "public_sink";
export const LABEL_DESTRUCTIVE = "destructive";
export const LABEL_UNTRUSTED = "untrusted_content";
export const LABEL_PRIVATE = "private_data";

// ═══════════════════════════════════════════════════════════════════════
// Known Server Classifications
// ═══════════════════════════════════════════════════════════════════════

export const KNOWN_SERVER_LABELS: Record<string, Set<string>> = {
  filesystem: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  fs: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  slack: new Set([LABEL_PUBLIC_SINK]),
  discord: new Set([LABEL_PUBLIC_SINK]),
  email: new Set([LABEL_PUBLIC_SINK]),
  gmail: new Set([LABEL_PUBLIC_SINK]),
  smtp: new Set([LABEL_PUBLIC_SINK]),
  sendgrid: new Set([LABEL_PUBLIC_SINK]),
  twilio: new Set([LABEL_PUBLIC_SINK]),
  telegram: new Set([LABEL_PUBLIC_SINK]),
  teams: new Set([LABEL_PUBLIC_SINK]),
  webhook: new Set([LABEL_PUBLIC_SINK]),
  github: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  gitlab: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  bitbucket: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  linear: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  jira: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  notion: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  asana: new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE]),
  postgres: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  postgresql: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  mysql: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  sqlite: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  mongo: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  mongodb: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  redis: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE]),
  supabase: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK]),
  fetch: new Set([LABEL_UNTRUSTED]),
  puppeteer: new Set([LABEL_UNTRUSTED]),
  playwright: new Set([LABEL_UNTRUSTED]),
  browser: new Set([LABEL_UNTRUSTED]),
  "brave-search": new Set([LABEL_UNTRUSTED]),
  tavily: new Set([LABEL_UNTRUSTED]),
  "web-search": new Set([LABEL_UNTRUSTED]),
  scraper: new Set([LABEL_UNTRUSTED]),
  crawl: new Set([LABEL_UNTRUSTED]),
  aws: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK]),
  gcp: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK]),
  azure: new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE, LABEL_PUBLIC_SINK]),
  docker: new Set([LABEL_DESTRUCTIVE]),
  kubernetes: new Set([LABEL_DESTRUCTIVE]),
  k8s: new Set([LABEL_DESTRUCTIVE]),
  terraform: new Set([LABEL_DESTRUCTIVE]),
  shell: new Set([LABEL_DESTRUCTIVE, LABEL_UNTRUSTED]),
  terminal: new Set([LABEL_DESTRUCTIVE, LABEL_UNTRUSTED]),
  exec: new Set([LABEL_DESTRUCTIVE]),
  "code-runner": new Set([LABEL_DESTRUCTIVE]),
  sandbox: new Set([LABEL_DESTRUCTIVE]),
  memory: new Set([LABEL_PRIVATE]),
  knowledge: new Set([LABEL_PRIVATE]),
  vector: new Set([LABEL_PRIVATE]),
  sentry: new Set([LABEL_PRIVATE]),
  datadog: new Set([LABEL_PRIVATE]),
  grafana: new Set([LABEL_PRIVATE]),
  s3: new Set([LABEL_PRIVATE, LABEL_PUBLIC_SINK, LABEL_DESTRUCTIVE]),
  gcs: new Set([LABEL_PRIVATE, LABEL_PUBLIC_SINK, LABEL_DESTRUCTIVE]),
  drive: new Set([LABEL_PRIVATE, LABEL_PUBLIC_SINK]),
  dropbox: new Set([LABEL_PRIVATE, LABEL_PUBLIC_SINK]),
};

const NAME_HEURISTICS: Array<[RegExp, Set<string>]> = [
  [/(?:file|fs|disk)/i, new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE])],
  [/(?:mail|email|smtp)/i, new Set([LABEL_PUBLIC_SINK])],
  [/(?:http|fetch|web|browser|scrape|crawl)/i, new Set([LABEL_UNTRUSTED])],
  [/(?:db|sql|database|mongo|redis)/i, new Set([LABEL_PRIVATE])],
  [/(?:exec|shell|command|terminal|run)/i, new Set([LABEL_DESTRUCTIVE])],
  [/(?:slack|discord|teams|telegram|chat)/i, new Set([LABEL_PUBLIC_SINK])],
  [/(?:github|gitlab|bitbucket|jira|linear)/i, new Set([LABEL_PUBLIC_SINK, LABEL_PRIVATE])],
  [/(?:aws|gcp|azure|cloud)/i, new Set([LABEL_PRIVATE, LABEL_DESTRUCTIVE])],
  [/(?:docker|k8s|kubernetes|terraform)/i, new Set([LABEL_DESTRUCTIVE])],
  [/(?:s3|gcs|storage|drive|dropbox)/i, new Set([LABEL_PRIVATE, LABEL_PUBLIC_SINK])],
];

// ═══════════════════════════════════════════════════════════════════════
// Server Classification
// ═══════════════════════════════════════════════════════════════════════

export function classifyServer(server: Record<string, any>): Set<string> {
  const name = (server.name ?? "").toLowerCase().trim();
  const command = (server.command ?? "").toLowerCase();
  const argsStr = (server.args ?? [])
    .filter((a: any): a is string => typeof a === "string")
    .join(" ")
    .toLowerCase();

  // Exact match
  if (KNOWN_SERVER_LABELS[name]) {
    return new Set(KNOWN_SERVER_LABELS[name]);
  }

  // Substring match in name
  for (const [known, labels] of Object.entries(KNOWN_SERVER_LABELS)) {
    if (name.includes(known)) return new Set(labels);
  }

  // Substring match in command/args
  const searchText = `${command} ${argsStr}`;
  for (const [known, labels] of Object.entries(KNOWN_SERVER_LABELS)) {
    if (searchText.includes(known)) return new Set(labels);
  }

  // Heuristic patterns
  const labels = new Set<string>();
  for (const [pattern, hLabels] of NAME_HEURISTICS) {
    if (pattern.test(name) || pattern.test(command) || pattern.test(argsStr)) {
      for (const l of hLabels) labels.add(l);
    }
  }

  return labels;
}

// ═══════════════════════════════════════════════════════════════════════
// Dangerous Combination Detection
// ═══════════════════════════════════════════════════════════════════════

function detectCombos(serverLabels: Map<string, Set<string>>): ToxicFlowResult[] {
  const flows: ToxicFlowResult[] = [];

  const allLabels = new Set<string>();
  for (const labels of serverLabels.values()) {
    for (const l of labels) allLabels.add(l);
  }

  // Servers by label
  const byLabel = new Map<string, string[]>();
  for (const [name, labels] of serverLabels) {
    for (const label of labels) {
      if (!byLabel.has(label)) byLabel.set(label, []);
      byLabel.get(label)!.push(name);
    }
  }

  const has = (l: string) => allLabels.has(l);
  const serversFor = (...labels: string[]): string[] =>
    [...new Set(labels.flatMap((l) => byLabel.get(l) ?? []))].sort();

  // Full chain: untrusted + private + sink
  if (has(LABEL_UNTRUSTED) && has(LABEL_PRIVATE) && has(LABEL_PUBLIC_SINK)) {
    flows.push({
      risk_level: "high",
      risk_type: "full_chain",
      title: "Full attack chain detected",
      description:
        "This agent can fetch external content, read private data, " +
        "and send data externally. An attacker could inject instructions " +
        "via fetched content, read sensitive files, and exfiltrate them.",
      servers_involved: serversFor(LABEL_UNTRUSTED, LABEL_PRIVATE, LABEL_PUBLIC_SINK),
      labels_involved: [LABEL_UNTRUSTED, LABEL_PRIVATE, LABEL_PUBLIC_SINK],
      remediation:
        "Scope filesystem access to non-sensitive directories. " +
        "Remove or restrict external communication servers.",
      tools_involved: [],
    });
    return flows; // Full chain subsumes individual combos
  }

  // Data exfiltration: private + sink
  if (has(LABEL_PRIVATE) && has(LABEL_PUBLIC_SINK)) {
    flows.push({
      risk_level: "high",
      risk_type: "data_exfiltration",
      title: "Data exfiltration path detected",
      description:
        "This agent can read private data and send it externally. " +
        "A prompt injection could instruct the agent to read sensitive " +
        "files and leak them via an external service.",
      servers_involved: serversFor(LABEL_PRIVATE, LABEL_PUBLIC_SINK),
      labels_involved: [LABEL_PRIVATE, LABEL_PUBLIC_SINK],
      remediation:
        "Scope filesystem access to non-sensitive directories only. " +
        "Review which external services truly need write access.",
      tools_involved: [],
    });
  }

  // Remote code execution: untrusted + destructive
  if (has(LABEL_UNTRUSTED) && has(LABEL_DESTRUCTIVE)) {
    flows.push({
      risk_level: "high",
      risk_type: "remote_code_execution",
      title: "Remote code execution path detected",
      description:
        "This agent can fetch external content and execute destructive " +
        "operations. Fetched content could contain malicious instructions " +
        "that modify files, execute commands, or alter databases.",
      servers_involved: serversFor(LABEL_UNTRUSTED, LABEL_DESTRUCTIVE),
      labels_involved: [LABEL_UNTRUSTED, LABEL_DESTRUCTIVE],
      remediation:
        "Add confirmation steps before destructive operations. " +
        "Restrict or sandbox the execution server.",
      tools_involved: [],
    });
  }

  // Data destruction: private + destructive from different servers
  if (has(LABEL_PRIVATE) && has(LABEL_DESTRUCTIVE)) {
    const privateServers = new Set(byLabel.get(LABEL_PRIVATE) ?? []);
    const destructiveServers = new Set(byLabel.get(LABEL_DESTRUCTIVE) ?? []);
    const same = privateServers.size === destructiveServers.size &&
      [...privateServers].every((s) => destructiveServers.has(s));
    if (!same) {
      flows.push({
        risk_level: "medium",
        risk_type: "data_destruction",
        title: "Data destruction path detected",
        description:
          "This agent can read private data from one source and " +
          "perform destructive operations on another. This could " +
          "lead to data corruption or deletion.",
        servers_involved: [...new Set([...privateServers, ...destructiveServers])].sort(),
        labels_involved: [LABEL_PRIVATE, LABEL_DESTRUCTIVE],
        remediation:
          "Review whether both data read and write capabilities " +
          "are necessary. Consider read-only access where possible.",
        tools_involved: [],
      });
    }
  }

  return flows;
}

/**
 * Analyze MCP servers for dangerous capability combinations.
 * Requires at least 2 servers for cross-server flow detection.
 */
export function analyzeToxicFlows(servers: Array<Record<string, any>>): ToxicFlowResult[] {
  if (servers.length < 2) return [];

  const serverLabels = new Map<string, Set<string>>();
  for (const srv of servers) {
    const name: string = srv.name ?? "unknown";
    const labels = classifyServer(srv);
    if (labels.size > 0) {
      serverLabels.set(name, labels);
    }
  }

  if (serverLabels.size === 0) return [];

  return detectCombos(serverLabels);
}
